### PR TITLE
Cis2 invoke

### LINKED ISSUE #71: Add support for CIS2 specific invoke transactions to the wallet-proxy
**Task description**

Add support for CIS2 tokens to the wallet-proxy.

The mobile wallet needs to access the following token information
- [ ] A list of token metadata URLs for each token.
- [ ] A balance of a token for contract + token id + account address.

### PR DESCRIPTION
**Pull request by abizjak**
## Purpose

Closes #71 

## Changes

Add two new endpoints to handle the required requests.

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [ ] I have performed a self-review of the changes.
- [ ] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [ ] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -29,7 +29,10 @@ The wallet proxy provides the following endpoints:
 * `GET /v0/appSettings`: get the up-to-date status of the old mobile wallet app.
 * `GET /v1/appSettings`: get the up-to-date status of the new mobile wallet app.
 * `GET /v0/epochLength`: get the epoch length in milliseconds.
-* `GET /v0/CIS2Tokens`: get the list of tokens on a given contract address.
+* `GET /v0/CIS2Tokens/{index}/{subindex}`: get the list of tokens on a given contract address.
+* `GET /v0/CIS2TokenMetadata/{index}/{subindex}`: get the metadata of tokens in on given contract address.
+* `GET /v0/CIS2TokenMetadata/{index}/{subindex}/{account address}`: get the metadata of tokens in on given contract address.
```
Comment (limemloh) [inline README.md:34]:
```suggestion
* `GET /v0/CIS2TokenBalance/{index}/{subindex}/{account address}`: get the balance of tokens in on given contract address for a given account address.
```

File: README.md [hunk 2]
```diff
@@ -729,6 +732,92 @@ The return value is an object with fields
...
+- `tokenId`: a comma separated list of token IDs. Token IDs are hex encoded, in
+  the same format as that returned by `CIS2Tokens` endpoint. An empty string is interpreted
+  as a single token with an empty ID.
+
+The return value is a JSON list of objects with fields
+
+- `balance` (required) ... a string that contains the balance as a decimal
+  number. The balance is always non-negative, but can be very large, and will
+  not fit into a 64-bit integer in general, so unbounded integer types should be
+  used.
```
Comment (limemloh) [inline README.md:794]: The CIS-2 standard limits it to 256 bits

Comment (limemloh) [inline README.md:794]: Maybe also mention that the metadata contains the info of how to display this number. i.e how many decimals

File: src/Proxy.hs [hunk 3]
```diff
@@ -1014,6 +1026,184 @@ getCIS2Tokens index subindex = do
...
+            tids <- parseCIS2TokenIds
+            let serializedParam = Wasm.Parameter . BSS.toShort . S.runPut $ do
+                    S.putWord16le (fromIntegral (length tids))
+                    mapM_ S.put (zip tids (repeat (AddressAccount addr)))
+            let contractAddr = ContractAddress (ContractIndex index) (ContractSubindex subindex)
+            cis2InvokeHelper contractAddr (Wasm.EntrypointName "balanceOf") serializedParam nrg $ \rv -> do
+                let getURLs = do
+                        len <- S.getWord16le
+                        replicateM (fromIntegral len) getTokenBalance
+                case S.runGet getURLs rv of
```
Comment (limemloh) [inline src/Proxy.hs:1093]:
```suggestion
                let getBalances = do
                        len <- S.getWord16le
                        replicateM (fromIntegral len) getTokenBalance
                case S.runGet getBalances rv of
```

**Review by limemloh — APPROVED** [review]
Looks good.
No major comments

**Review by shjortConcordium — APPROVED** [review]
LGTM (considering Emil's comments :upside_down_face: )

File: README.md (hunk 2 above)
Comment (abizjak) [reply README.md:794]: Done

**Pull request merged**